sound/soc/sof/intel/telemetry.c

```diff
@@ -0,0 +1,87 @@
+// SPDX-License-Identifier: (GPL-2.0-only OR BSD-3-Clause)
+//
+// This file is provided under a dual BSD/GPLv2 license.  When using or
+// redistributing this file, you may do so under either license.
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+//
+
+#include <sound/sof/ipc4/header.h>
+#include <sound/sof/xtensa.h>
+#include "../ipc4-priv.h"
+#include "../sof-priv.h"
+#include "hda.h"
+#include "telemetry.h"
+
+void sof_ipc4_intel_dump_telemetry_state(struct snd_sof_dev *sdev, u32 flags)
+{
+	struct telemetry_slot_data *telemetry_data;
+	struct sof_ipc_dsp_oops_xtensa *xoops;
+	u32 slot_offset;
+	char *level;
+
+	slot_offset = sof_ipc4_find_debug_slot_offset_by_type(sdev, SOF_IPC4_DEBUG_SLOT_TELEMETRY);
+	if (!slot_offset)
+		return;
+
+	telemetry_data = kmalloc(sizeof(*telemetry_data), GFP_KERNEL);
+	if (!telemetry_data)
+		return;
+
+	sof_mailbox_read(sdev, slot_offset, telemetry_data, sizeof(*telemetry_data));
+	if (telemetry_data->separator != XTENSA_CORE_DUMP_SEPARATOR) {
+		dev_err(sdev->dev, "Core dump information is not available due to invalid separator %#x\n",
+			telemetry_data->separator);
+		goto free;
+	}
```

Comment:
I would get this under the acceptable character count by:
```c
	struct device *dev = sdev->dev;
	static const char invalid_slot_msg[] = "Core dump is not available due to";

...

	if (telemetry_data->hdr.id[0] != COREDUMP_HDR_ID0 ||
	    telemetry_data->hdr.id[1] != COREDUMP_HDR_ID1 ||
	    telemetry_data->arch_hdr.id != COREDUMP_ARCH_HDR_ID) {
		dev_err(dev, "%s invalid coredump header %c%c, arch hdr %c\n",
			invalid_msg, telemetry_data->hdr.id[0],
			telemetry_data->hdr.id[1],
			telemetry_data->arch_hdr.id);
		goto free;
	}
```
it might hurt grepabitlity for the whole sentence, but this just ends at 99:
```c
	if (telemetry_data->hdr.id[0] != COREDUMP_HDR_ID0 ||
	    telemetry_data->hdr.id[1] != COREDUMP_HDR_ID1 ||
	    telemetry_data->arch_hdr.id != COREDUMP_ARCH_HDR_ID) {
		dev_err(sdev->dev,
			"Core dump is not available: invalid coredump header %c%c, arch hdr %c\n",
			telemetry_data->hdr.id[0], telemetry_data->hdr.id[1],
			telemetry_data->arch_hdr.id);
		goto free;
	}
```

Reply:
thanks, updatged